### ISSUE TITLE
[NNC] NNC type promotion logic adds extra casts & does compute in different dtype than eager

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#49178]
## 🐛 Bug

Almost all NNC type promotion goes through a common path which [promotes inputs according to the highest input type](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp#L725), and then after compute, casts the output to whichever type was recorded. 

Because we have the output type specified, all NNC ops will give the correct dtype. However, the compute may be done in a different dtype than eager, and there may be extraneous casts added. 

In the (worst) case of something like, 
```
torch.addcmul(torch.ones([4,  4]), torch.ones([4,  4]), torch.ones([4,  4]), value =.5)
```

We will cast all three tensor inputs to floating point, do the addcmul, then cast the operation back. As opposed to eager, which will just cast `.5` to int. 

The advantage of the current approach is that we are guaranteed to have the same output dtype. I think ops could gradually be migrated over to have the correct casting behavior.

TODO: 

- [ ] addcmul

## To Reproduce

1. Comment out the [output type casting](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp#L728)  of the various `comput{number}Operands` 

2. Run `python test/test_jit_fuser_te..py`

**Comment by eellison** [comment]
Adding bootcamp here because each individual op is pretty easy to migrate over to the right casting logic. Starting with addcmul but we can add others.
